test/e2e/operator_test.go

```diff
@@ -228,3 +233,402 @@ func setImage(deployment *appsv1.Deployment, image string) {
 		}
 	}
 }
+
+func TestDNSForwarding(t *testing.T) {
+	cl, err := getClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Ensure the DNS Operator is available.
+	dns := &operatorv1.DNS{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Name: "default"}, dns); err != nil {
+		t.Fatalf("failed to get default dns: %v", err)
+	}
+
+	// Create the upstream resolver ConfigMap.
+	upstreamPodName := "test-upstream"
+	upstreamPodNS := "openshift-dns"
+	upstreamCfgMap := upstreamConfigMap(upstreamPodName, upstreamPodNS)
+	if err := cl.Create(context.TODO(), upstreamCfgMap); err != nil {
+		t.Fatalf("failed to create configmap %s/%s: %v", upstreamCfgMap.Namespace, upstreamCfgMap.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamCfgMap); err != nil {
+			t.Fatalf("failed to delete configmap %s/%s: %v", upstreamCfgMap.Namespace, upstreamCfgMap.Name, err)
+		}
+	}()
+
+	// Get the CoreDNS image used by the test upstream resolver.
+	co := &configv1.ClusterOperator{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Name: operatorcontroller.DNSOperatorName}, co); err != nil {
+		t.Fatalf("failed to get cluster operator dns: %v", err)
+	}
+	var upstreamImage string
+	for _, ver := range co.Status.Versions {
+		if ver.Name == "coredns" {
+			if len(ver.Version) == 0 {
+				t.Fatalf("dns operator does not contain a coredns version: %v", err)
+			}
+			upstreamImage = ver.Version
+			break
+		}
+	}
+
+	// Create the upstream resolver Pod.
+	upstreamResolver := upstreamPod(upstreamPodName, upstreamPodNS, upstreamImage, upstreamPodName)
+	if err := cl.Create(context.TODO(), upstreamResolver); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamResolver); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+		}
+	}()
+
+	// Wait for the upstream resolver Pod to be ready.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: upstreamResolver.Namespace, Name: upstreamResolver.Name}, upstreamResolver); err != nil {
+			return false, nil
+		}
+		for _, cond := range upstreamResolver.Status.Conditions {
+			if cond.Type == corev1.ContainersReady &&
+				cond.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe ContainersReady condition for pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+
+	// Create the upstream resolver Service and get the ClusterIP.
```

Comment:
I wonder if you even need a service? Could possibly address the pod directly via IP or pod DNS name? Could help reduce the amount of work to do

Reply:
> I wonder if you even need a service? Could possibly address the pod directly via IP or pod DNS name?

@ironcladlou we can, but I was trying to replicate the default DNS deployment as much as possible. Let me know if you prefer I remove the Service layer.